crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -111,46 +128,93 @@ impl CoinGecko {
             })?
             .0;
 
-        let price = prices
+        Ok(tokens
+            .iter()
+            .map(|token| {
+                (
+                    **token,
+                    prices
+                        .get(token)
+                        .ok_or(PriceEstimationError::NoLiquidity)
+                        .map(|price| price.eth),
+                )
+            })
+            .collect())
+    }
+
+    pub async fn send_request_price_in_eth(&self, token: Token) -> NativePriceEstimateResult {
+        let prices = self.send_bulk_request_price_in_eth(&[&token]).await?;
+        prices
             .get(&token)
-            .ok_or(PriceEstimationError::NoLiquidity)?;
-        Ok(price.eth)
+            .ok_or(PriceEstimationError::NoLiquidity)?
+            .clone()
+    }
+}
+
+#[async_trait]
+impl NativePriceBatchFetcher for CoinGecko {
+    async fn fetch_native_prices(
+        &self,
+        tokens: &HashSet<H160>,
+    ) -> std::result::Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError> {
+        let mut requested_tokens = tokens.iter().collect::<Vec<_>>();
+        match self.quote_token {
+            QuoteToken::Eth => self.send_bulk_request_price_in_eth(&requested_tokens).await,
+            QuoteToken::Other(native_price_token) => {
+                requested_tokens.push(&native_price_token);
```

Comment:
Any chance `native_price_token` already exists in the `requested_tokens`? No issue with sending requests with duplicated tokens?

Reply:
It is not a problem to be twice, but actually there may be an issue with the limit. Fixing it. Thank you!

Reply:
So, does the batch size of the batched transport, in fact, control this limit? Say the coin gecko limit is 50. Should the batch size also take into account any additional tokens that might be added to the request?

Reply:
Yup, that's exactly the issue I am fixing. Right now the limit is enforced by the PR 1/3 (https://github.com/cowprotocol/services/pull/2869), but this additional token may be a problem if the request already is already on its limit, so we should fire another request for the native price token.